tiny-event-sourcing-lib/src/main/kotlin/ru/quipy/streams/EventStoreReader.kt

```diff
@@ -0,0 +1,142 @@
+
+    /**
+     * We can "replay" events in the stream by resetting it to desired reading index
+     */
+    fun resetReadIndex(resetInfo: ReadIndexResetInfo)
+    fun stop()
+}
+
+class EventStoreReader(
+    private val eventStore: EventStore,
+    private val streamName: String,
+    private val tableName: String,
+    private val streamManager: EventStreamReaderManager,
+    private val eventStreamNotifier: EventStreamNotifier,
+    private val dispatcher: CoroutineDispatcher
+) : EventReader {
+    companion object {
+        private val NO_RESET_REQUIRED = ReadIndexResetInfo(-1)
+    }
+
+    private val logger: Logger = LoggerFactory.getLogger(EventStoreReader::class.java)
+    private val nextReaderAliveCheck: Duration = 15.seconds
+
+    private val isReaderPrimary: AtomicBoolean = AtomicBoolean(false)
+    private val isScanActive: Boolean = true
+
+    private val scanJob: Job = launchEventStoreReaderScanJob()
+
+    private var streamReadIndex: EventStreamReadIndex = EventStreamReadIndex(streamName, readIndex = 0L, version = 0L)
+    private var indexResetInfo: ReadIndexResetInfo = NO_RESET_REQUIRED
+    private var processedRecords = 0L
+
+    override suspend fun read(batchSize: Int): List<EventRecord> {
+        if (!isReaderPrimary.get())
+            return emptyList()
+
+        checkAndResetIndexIfRequired()
+
+        val eventRecords = eventStore.findBatchOfEventRecordAfter(tableName, streamReadIndex.readIndex, batchSize)
+        eventStreamNotifier.onBatchRead(streamName, eventRecords.size)
+
+        return eventRecords
+    }
+
+    override fun postProcessRecord(eventRecord: EventRecord) {
+        val processingRecordTimestamp = eventRecord.createdAt
+
+        streamReadIndex = EventStreamReadIndex(streamName, processingRecordTimestamp, streamReadIndex.version)
+
+        if (processedRecords++ % 10 == 0L)
+            commitReadIndex(processingRecordTimestamp)
+    }
+
+    override fun resetReadIndex(resetInfo: ReadIndexResetInfo) {
+        if (streamReadIndex.version < 1)
+            throw IllegalArgumentException("Can't reset to non existing version: ${streamReadIndex.version}")
+
+        indexResetInfo = resetInfo
+    }
+
+    override fun stop() {
+        if (isReaderPrimary.get())
+            isReaderPrimary.set(false)
+
+        if (isScanActive)
+            scanJob.cancel()
+    }
+
+    private fun commitReadIndex(index: Long) {
+        EventStreamReadIndex(streamName, index, streamReadIndex.version + 1L).also {
+            logger.info("Committing index for $streamName, index: ${index}, current version: ${streamReadIndex.version}")
+            eventStore.commitStreamReadIndex(it)
+            eventStreamNotifier.onReadIndexCommitted(streamName, it.readIndex)
+        }
+
+        syncReaderIndex()
+    }
+
+    private fun launchEventStoreReaderScanJob(): Job {
+        return CoroutineScope(CoroutineName("reading-$streamName-coroutine") + dispatcher).launch {
+            while (isScanActive) {
```

Comment:
maybe like "healthCheck", because scan is a bit confusing

Reply:
Done.